lib/compiler/stack.js

```diff
@@ -0,0 +1,179 @@
+"use strict";
+
+class Stack {
+  constructor(ruleName, varName, type) {
+    /** Last used variable in the stack. */
+    this.sp       = -1;
+    /** Maximum stack size. */
+    this.maxSp    = -1;
+    this.varName  = varName;
+    this.ruleName = ruleName;
+    this.type     = type;
+  }
+
+  /**
+   * Returns name of the variable at the index `i`.
+   *
+   * @param {number} i Index for which name must be generated
+   * @return {string} Generated name
+   *
+   * @throws {RangeError} If `i < 0` which means stack underflow (there a more `pop`'s than `push`'s)
+   */
+  name(i) {
+    if (i < 0) {
+      throw new RangeError(
+        `Rule '${this.ruleName}': Var stack underflow: attempt to use a variable '${this.varName}<x>' at index ${i}`
+      );
+    }
+
+    return this.varName + i;
+  }
+
+  /**
+   * Assigns `exprCode` to the new variable in the stack, returns generated code.
+   * As the result, the size of a stack increases on 1.
+   *
+   * @param {string} exprCode Any expression code that must be assigned to the new variable in the stack
+   * @return {string} Assignment code
+   */
+  push(exprCode) {
+    const code = this.name(++this.sp) + " = " + exprCode + ";";
```

Comment:
Even though this is only used from generate-js at the moment, I was surprised to find JS code here.  a) it makes it harder to see all of the JS code in one place b) it makes it harder to use this class for other languages later.

If you decide to do something about this, then `this.type` may or may not still be needed in the class.

Reply:
Actually, this class can be used by almost any C-like code generator: C/C++/C#/JS/Java. I'm not consider it as a public API right now, although

Reply:
OK, that makes sense.  Later, if we want to be able to generate languages like Python, we can deal with `type` and semicolon being optional without breaking any code that relies on this interface.